Prevent user from jailbreak #289

gdaguin opened the pull request:
Prevent user from jailbreak.

# Description

The first time we detect that the device is unsecure (jailbroken or dev mode), we will warn the user with a popup. Next time he launches the application, the user is going to see a snackbar. Creation of a security manager to have the capability to know the device security (jailbreak or dev mode right now) from every where in the app

Fixes #289 

## Type of change
- New feature (non-breaking change which adds functionality)

## Material used:

Please delete options that are not relevant.
- iOS (Smartphone)
- Android (Smartphone)

## How Has This Been Tested?

I tested in an iOS sim (dev mode) and an Android device in dev mode and with no dev mode. I tested the first appearance of the popup and then the appearance of the snackbar. 

### Patrol
-

## Checklist:
- My code follows the style guidelines of this project
- I have performed a self-review of my own code
- I have commented my code, particularly in hard-to-understand areas
- I have made corresponding changes to the documentation
- My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- Any dependent changes have been merged and published in downstream modules

## Useful info for the reviewer:

Commit:
Fix issue #289

Prevent user from jailbreak. The first time we detect that the device is unsecure (jailbroken or dev mode), we will warn the user with a popup. Next time he launches the application, the user is going to see a snackbar.
Creation of a security manager to have the capability to know the device security (jailbreak or dev mode right now) from every where in the app

gdaguin commented on lib/main.dart:
```diff
@@ -254,36 +253,6 @@ class SplashState extends ConsumerState<Splash> with WidgetsBindingObserver {
-            cancelText: AppLocalizations.of(context)!.exit.toUpperCase(),
-            cancelAction: () {
-              if (!kIsWeb && Platform.isIOS) {
-                exit(0);
```
I didn't manage the yes/no possibility within the popup because Apple can refuse the publication of the application if we "exit" the app : https://developer.apple.com/forums/thread/672098

gdaguin commented on lib/main.dart:
```diff
@@ -302,32 +271,75 @@ class SplashState extends ConsumerState<Splash> with WidgetsBindingObserver {
-  Future<void> _goToIntroScreen() async {
-    await Navigator.of(context).pushReplacementNamed('/intro_welcome');
+  void _goToIntroScreen() {
+    Navigator.of(context).pushReplacementNamed('/intro_welcome');
```
synchronous method now because the await will end when the page "intro_welcome" has been pushed in the navigation stack and is not in the top anymore. 
I add some code after the call of checkLoggedIn (that calls _goToIntroScreen), that's why I needed to change that.

redDwarf03 commented on lib/main.dart:
```diff
@@ -302,32 +271,75 @@ class SplashState extends ConsumerState<Splash> with WidgetsBindingObserver {
+      return;
+    }
+
+    final preferences = await HivePreferencesDatasource.getInstance();
```
Perhaps we could use a provider and watch it to check Device Security because if user leavexs a wallet open and changes the developer mode for example, he won't be informed

gdaguin replied:
Nice one, I hadn't thought of that ! No need to use a provider to do that I think, I will check what I can do.

Commit:
Lines length

Commit:
Support for changing developer or jailbreak mode when app is in background

Also fixed an issue with the app translation update when the device language changes

gdaguin commented:
When the app is back in foreground, it checks whether or not the dev mode (or jailbreak) has been updated since the last time the app entered in background. I think this is pretty accurate right now.

Chralu commented on lib/main.dart:
```diff
@@ -115,12 +119,75 @@ Future<void> main() async {
+    didChangeAppLifecycleStateAsync(state);
+  }
+
+  void didChangeAppLifecycleStateAsync(AppLifecycleState state) async {
```
Why moving those behaviors from the SplashScreen ?

gdaguin replied:
Because it needs to be aware of the app state from whatever screen and not only the splash screen. I did some tests and I noticed a bug that once the Splash screen disappears, the previous methods don't work anymore because the splash screen is not in the navigation stack anymore.

redDwarf03 commented on lib/main.dart:
Please remove unused imports

redDwarf03 commented on lib/main.dart:
```diff
@@ -115,12 +119,75 @@ Future<void> main() async {
+    didChangeAppLifecycleStateAsync(state);
+  }
+
+  void didChangeAppLifecycleStateAsync(AppLifecycleState state) async {
```
This method should return a Future<void>

redDwarf03 commented on lib/util/security_manager.dart:
```diff
@@ -0,0 +1,79 @@
+  }
+
+  Future checkDeviceSecurity(WidgetRef ref, BuildContext context) async {
+    if (Platform.isIOS == false && Platform.isAndroid == false) {
```
Warning, the wallet should work on web platform too.
Could you add kIsWeb condition please

gdaguin replied:
We are only checking if the code is running in iOS or Android, to know if the device is jailbroken or in dev mode, we don't want to test if it is running in web in that case, right ?

redDwarf03 replied:
yes. but an error occurs on ```io.Platform```class when we execute the wallet in web mode.
So protect the condition with ```!kIsWeb && (Platform....```

redDwarf03 reviewed (changes requested):
Except minors changes, this PR is ok for acceptance tests campaign for the next version 2.1.1. could you please manage the request and rebase from dev too 
Thank you

redDwarf03 commented on lib/infrastructure/datasources/hive_preferences.dart:
```diff
@@ -64,6 +64,9 @@ class HivePreferencesDatasource {
 
   Future<void> setHasSeenRootWarning() => _setValue(hasShownRootWarning, true);
 
+  Future<void> setHasNotSeenRootWarning() =>
```
Why don't you use a single setter method with a bool in param ?

Commit:
MR updates

gdaguin commented on lib/util/security_manager.dart:
```diff
@@ -0,0 +1,83 @@
+  }
+
+  Future checkDeviceSecurity(WidgetRef ref, BuildContext context) async {
+    if (kIsWeb == true) {
```
I hate || so I used two if instead

gdaguin committed:
Merge branch 'archethic-foundation:dev' into issue289
